server/src/main/java/org/elasticsearch/common/time/DateUtils.java

```diff
@@ -129,6 +129,27 @@ public static Instant toInstant(long nanoSecondsSinceEpoch) {
         return Instant.ofEpochSecond(seconds, nanos);
     }
 
+    /**
+     * Convert a nanosecond timestamp in milliseconds
+     *
+     * @param milliSecondsSinceEpoch the millisecond since the epoch
+     * @return                      the nanoseconds since the epoch
+     */
+    public static long toNanoSeconds(long milliSecondsSinceEpoch) {
+        if (milliSecondsSinceEpoch < 0) {
+            throw new IllegalArgumentException("milliSeconds [" + milliSecondsSinceEpoch + "] are before the epoch in 1970 and cannot " +
+                "be converted to nanoseconds");
+        } else if (milliSecondsSinceEpoch > MAX_NANOSECOND_INSTANT.toEpochMilli()) {
```

Comment:
how costly is the toEpochMillis call?

Reply:
Yes it's costly to call it on every value, I pushed https://github.com/elastic/elasticsearch/pull/40199/commits/b0e2a2007330ecadacce0e2a1ccfa623f69b7717 to compute this value once.